CSharp/JobPrep/Program.cs

```diff
@@ -0,0 +1,333 @@
+                // Using optional select clause to return only the name and state.
+                IEnumerable<ICloudPool> pools = pm.ListPools(new ODATADetailLevel(selectClause: "name,state"));
+                foreach (var p in pools)
+                {
+                    Console.WriteLine("pool " + p.Name + " is " + p.State);
+                }
+                Console.WriteLine();
+            }
+        }
+
+        private static void ListWorkItems(IBatchClient client)
+        {
+            // All Workitem, Job, and Task related operation start from WorkItemManager
+            using (IWorkItemManager wm = client.OpenWorkItemManager())
+            {
+                Console.WriteLine("Listing Workitems\n=================");
+                IEnumerable<ICloudWorkItem> wis = wm.ListWorkItems();
+                foreach (var w in wis)
+                {
+                    Console.WriteLine("Workitem: " + w.Name + " State:" + w.State);
+                }
+                Console.WriteLine();
+            }
+        }
+
+        private static async Task<bool> PollJobPrepTaskAsync(IPoolManager pm, string pool, string vm, string workitem, string job, int sleep)
+        {
+            while(true)
+            {
+                try
+                {
+                    Console.WriteLine("polling prep task on " + vm);
+                    var preptask = await pm.GetJobPreparationTaskStatusAsync(pool, vm, workitem, job);
+                    if (preptask.State == JobPreparationTaskState.Completed)
+                        return preptask.ExitCode == 0;
+                    Console.WriteLine("sleep a sec");
+                    await Task.Delay(sleep);
+                }
+                catch (AggregateException) { // ignore not found exception
+                    Console.WriteLine("caught exception in async method");
+                }
+            }
+        }
+
+        private static void Log(string msg, string vm = "")
+        {
+#if VERBOSE
+            Console.WriteLine(DateTime.Now.ToString("HH:mm:ss.ffff ") + vm + " " + msg);
+#endif
+        }
+
+        private static bool WaitAllJobPrepTask(IBatchClient client, string pool, string workitem, string job, Func<int, int, bool> stop, Func<int, int, bool> succeed, int timeout = 60*1000, int freq = 2000, int parallelism = 5)
+        {
+            using (IPoolManager pm = client.OpenPoolManager())
+            {
+                using (IWorkItemManager wm = client.OpenWorkItemManager())
+                {
+
+                    CancellationTokenSource cts = new CancellationTokenSource();
+                    CancellationToken ct = cts.Token;
+
+                    BlockingCollection<string> vms = new BlockingCollection<string>();
+                    List<Task> tasks = new List<Task>();
+                    int nb_succeeded = 0;
+                    int nb_failed = 0;
+
+                    // find all VM in the Pool and put them in a queue
+                    Log("add vms");
```

Comment:
This line is confusing because you aren't adding VMs to the service, you're just adding them to a list... you should clarify that.